### ISSUE TITLE
Can class fields be assigned in pattern assignments?

### ISSUE THREAD

**Issue by deakjahn** [dart-lang/language#3096]
Trying to decompose a record to class fields gives the error "Only local variables can be assigned in pattern assignments". Is this unavoidable? Couldn't it be solved so that we can assign to class fields as well?

**Comment by deakjahn** [comment]
As a first step, the "plain identifiers" restriction would seem acceptable to me but not only to local variables. Straight fields, in my opinion, might be needed in this place very often.

**Comment by munificent** [comment]
Even that restriction can get weird because you could still be invoking a setter method or initializing a `late` field. It's probably *tractable* and I can see us enabling this at some point. The initial version of the proposal did. But it does get sort of complex which is why we don't currently allow it.

**Comment by ilovelinux** [comment]
> As a first step, the "plain identifiers" restriction would seem acceptable to me but not only to local variables. Straight fields, in my opinion, might be needed in this place very often.

Totally agree. My use case was about assigning values to instance variables.

**Comment by natebosch** [comment]
I just ran into this today in a review. I wanted to suggest that a serialized set of async tasks could use the record `.wait` extension, but I can't do it when some of the assignments are to class fields.

**Comment by munificent** [comment]
Yeah, it's a reasonable extension, we just have to work through all of the consequences around `late` and lazy statics.

Also, if we allow irrefutable pattern assignment expressions to assign to fields, then users might also expect patterns in field and top-level declarations, like:

```dart
class C {
  final (int x, int y);
}
```

Then we have to figure out how that interacts with `late` and initializer lists.

**Comment by munificent** [comment]
I considered closing this as a duplicate of #2855, but I think I'll keep this open as a separate issue. #2855 is about allowing more assignable target syntaxes. This one is about having the same syntax but an identifier that resolves to something other than a local variable. I could see us doing one of these features without the other, so I'll keep both of these issues open.
